Determine any memory/CPU limitations from sysfs cgroup

nealef opened the pull request:
Add capability to interrogate cgroup limitations when determining CP and memory limits

This code has been adapted from coreCLR. It has been modified from C++ but uses the same naming conventions in the event of a unified mechanism that can be shared between both runtimes being developed. The code has been tested on Ubuntu 20.04 and CentOS 7 with cgroupv1 and cgroupv2.

This code is required in the event of running runtime in a container as the current limitations being discovered by the mono runtime are purely for the machine and not in a container which may have lower quotas.

* src/mono/CMakeLists.txt
  - Set the HAVE_CGROUP_SUPPORT for Linux hosts

* src/mono/cmake/config.h.in
  - Place holder for HAVE_CGROU_SUPPORT definition

* src/mono/mono/metadata/icall.c
* src/mono/mono/sgen/sgen-marksweep.c
* src/mono/mono/sgen/sgen-simple-nursery.c
  - Use mono_cpu_limit() instead of mono_cpu_count()

* src/mono/mono/utils/CMakeLists.txt
  - Add mono-cgroup.c to the build

* src/mono/mono/utils/memfuncs.c
  - Call `getRestrictedPhysicalMemoryLimit()` or `getPhyscalMemoryAvail()`

* src/mono/mono/utils/memfuncs.h
  - Add prototypes for the new APIs

* src/mono/mono/utils/mono-cgroup.c
  - Code adapted from coreCLR to interrogate sysfs to determine any limitations on memory or CPU

* src/mono/mono/utils/mono-proclib.c
  - Add call to `getCpuLimit()`

* src/mono/mono/utils/mono-proclib.h
  - Add prototype for the new API

nealef committed:
Determine any memory/CPU limitations from sysfs cgroup  (#21280)

Add capability to interrogate cgroup limitations when determining CP and memory limits

This code has been adapted from coreCLR. It has been modified from C++ but uses the same naming conventions in the event of a unified mechanism that can be shared between both runtimes being developed. The code has been tested on Ubuntu 20.04 and CentOS 7 with cgroupv1 and cgroupv2.

This code is required in the event of running runtime in a container as the current limitations being discovered by the mono runtime are purely for the machine and not in a container which may have lower quotas.

* src/mono/CMakeLists.txt
  - Set the HAVE_CGROUP_SUPPORT for Linux hosts

* src/mono/cmake/config.h.in
  - Place holder for HAVE_CGROU_SUPPORT definition

* src/mono/mono/metadata/icall.c
* src/mono/mono/sgen/sgen-marksweep.c
* src/mono/mono/sgen/sgen-simple-nursery.c
  - Use mono_cpu_limit() instead of mono_cpu_count()

* src/mono/mono/utils/CMakeLists.txt
  - Add mono-cgroup.c to the build

* src/mono/mono/utils/memfuncs.c
  - Call `getRestrictedPhysicalMemoryLimit()` or `getPhyscalMemoryAvail()`

* src/mono/mono/utils/memfuncs.h
  - Add prototypes for the new APIs

* src/mono/mono/utils/mono-cgroup.c
  - Code adapted from coreCLR to interrogate sysfs to determine any limitations on memory or CPU

* src/mono/mono/utils/mono-proclib.c
  - Add call to `getCpuLimit()`

* src/mono/mono/utils/mono-proclib.h
  - Add prototype for the new API

nealef committed:
Correct var type to match usage

naricc commented:
If this code is adapated from CoreCLR, is there a plan to share the code eventually, or need it all be duplicated?

akoeplinger commented on src/mono/mono/utils/memfuncs.c:
```diff
@@ -273,24 +274,59 @@ mono_determine_physical_ram_size (void)
+	if (restricted_limit != 0) {
+		gchar *heapHardLimit = getenv("DOTNET_GCHeapHardLimit");	/* See if user has set a limit */
+		if (heapHardLimit == NULL)
+			heapHardLimit = getenv("COMPlus_GCHeapHardLimit");	/* Check old envvar name */
```
the `DOTNET_` env var prefix is newer so I think it makes sense that it is preferred.

lambdageek commented on src/mono/mono/utils/mono-proclib.h:
```diff
@@ -19,5 +19,12 @@ MONO_API
 int
 mono_cpu_count (void);
 
+MONO_API 
+int
+mono_cpu_limit (void);
```
does this need to be `MONO_API`?
```suggestion
int
mono_cpu_limit (void);
```

akoeplinger replied:
pretty sure it does but @nealef would you mind checking?

nealef replied:
It is used in place of `mono_cpu_count()` in a couple of spots so I think it needs to have the same attributes.

lambdageek commented on src/mono/mono/utils/mono-cgroup.c:
```diff
@@ -0,0 +1,972 @@
+Abstract:
+    Read the memory limit for the current process
+
+    Adapted from runtime src/coreclr/gc/unix/cgroup.cpp 
```
Can we make a common version of this in `src/native` and share between the two runtimes?

nealef replied:
The coreCLR code is C++. So far as I can tell mono is entirely C. Is this something that can be done down the track? I am unsure how straightforward things would be.

lambdageek replied:
@nealef yea we woudl need to port the CoreCLR version to C (already done) and make C++ wrappers.  Similar to what was done for EventPipe.   I dont' think it needs to be part of this PR.  But we should make an issue to re-unify the code.

lambdageek commented on src/mono/mono/utils/mono-proclib.c:
```diff
@@ -190,3 +191,44 @@ mono_cpu_count (void)
+               char *dotnetProcCnt = getenv("DOTNET_PROCESSOR_COUNT");
+               if (dotnetProcCnt != NULL) {
+                       errno = 0;
+                       limit = strtol(dotnetProcCnt, NULL, 0);
```
implicit cast `long` to `int`. is that ok?

nealef replied:
I would use `atoi()` but there's no error checking. Do you want an explicit cast?

lambdageek replied:
well that's my question: should we just cast? do we need to check that the value's in range?

lambdageek commented on src/mono/mono/utils/mono-proclib.c:
```diff
@@ -190,3 +191,44 @@ mono_cpu_count (void)
+                       errno = 0;
+                       limit = strtol(dotnetProcCnt, NULL, 0);
+                       if ((errno == 0) && (limit > 0))        /* If it's in range and positive */
+                               return (limit);
```
```suggestion
                               return limit;
```

nealef replied:
Fixed

lambdageek commented on src/mono/mono/utils/mono-proclib.c:
```diff
@@ -190,3 +191,44 @@ mono_cpu_count (void)
+       /*
+        * Just return the cached value
+        */
+       return (limit);
```
```suggestion
       return limit;
```

nealef replied:
Fixed

lambdageek commented on src/mono/mono/utils/mono-proclib.h:
```diff
@@ -19,5 +19,12 @@ MONO_API
+mono_cpu_limit (void);
+
+gboolean 
+mono_get_cpu_limit(int *);
```
```suggestion
mono_get_cpu_limit(int *limit);
```

nealef replied:
Just a mono convention?

lambdageek commented on src/mono/mono/utils/memfuncs.c:
```diff
@@ -273,24 +274,59 @@ mono_determine_physical_ram_size (void)
+				restricted_limit = (3 * restricted_limit) / 4;	/* Use 75% limit of container */
+			}
+		}
+		return (restricted_limit < 209715200 ? 209715200 : 	/* Use at least 20MB */
```
where do these constants come from?

Can we use a `#define` for them?

nealef replied:
Done.

nealef commented on src/mono/mono/utils/mono-proclib.c:
```diff
@@ -190,3 +191,44 @@ mono_cpu_count (void)
+mono_cpu_limit (void)
+{
+#if defined(HOST_WIN32)
+       return mono_cpu_count();
```
Fixed

nealef committed:
Suggested updates

lambdageek reviewed (approved):
@SamMonoRT I think we want to take a backport of this for net7. I don't think this will impact mobile or wasm. (But the "DOTNET_PROCESSOR_COUNT" support might be a nice thing to bring along on those platforms for diagnostic scenarios)

SamMonoRT commented:
> @SamMonoRT I think we want to take a backport of this for net7. I don't think this will impact mobile or wasm. (But the "DOTNET_PROCESSOR_COUNT" support might be a nice thing to bring along on those platforms for diagnostic scenarios)

Yes, once all suggestions are addressed and CI lanes look green, will issue a backport command.

SamMonoRT commented:
@nealef - please let us know when all suggested changes are completed.

nealef commented:
The only one missing is the explicit cast from `strtol()`. I am having VPN problems today so I haven't been able to push. All other suggestions should have been pushed.

nealef committed:
* mono-cgroup.c
  - Adhere to call statement convention
  - Remove unnecessary cast
* mono-proclib.c
  - Explicit cast

akoeplinger commented on src/mono/mono/utils/mono-proclib.c:
```diff
@@ -190,3 +191,44 @@ mono_cpu_count (void)
+#if defined(HOST_WIN32)
+	return mono_cpu_count ();
+#else
+       static int limit = -1;  /* Value will be cached for future calls */
+
+       /*
+        * If 1st time through then check if user has mandated a value and use it,
+        * otherwise we check for any cgroup limit and use the min of actual number
+        * and that limit
+        */
+       if (limit == -1) {
+               char *dotnetProcCnt = getenv ("DOTNET_PROCESSOR_COUNT");
+               if (dotnetProcCnt != NULL) {
+                       errno = 0;
+                       limit = (int) strtol (dotnetProcCnt, NULL, 0);
+                       if ((errno == 0) && (limit > 0))        /* If it's in range and positive */
+                               return limit;
+               }
+               limit = mono_cpu_count ();
+#if HAVE_CGROUP_SUPPORT
+               int count = 0;
+               if (mono_get_cpu_limit (&count))
+                       limit = (limit < count ? limit : count);
+#endif
+       }
+
+       /*
+        * Just return the cached value
+        */
+       return limit;
+#endif
```
tabs instead of spaces:
```suggestion
	static int limit = -1;  /* Value will be cached for future calls */

	/*
	* If 1st time through then check if user has mandated a value and use it,
	* otherwise we check for any cgroup limit and use the min of actual number
	* and that limit
	*/
	if (limit == -1) {
		char *dotnetProcCnt = getenv ("DOTNET_PROCESSOR_COUNT");
		if (dotnetProcCnt != NULL) {
			errno = 0;
			limit = (int) strtol (dotnetProcCnt, NULL, 0);
			if ((errno == 0) && (limit > 0))        /* If it's in range and positive */
				return limit;
		}
		limit = mono_cpu_count ();
#if HAVE_CGROUP_SUPPORT
		int count = 0;
		if (mono_get_cpu_limit (&count))
			limit = (limit < count ? limit : count);
#endif
	}

	/*
	* Just return the cached value
	*/
	return limit;
#endif
```

nealef committed:
Convert spaces to tabs

akoeplinger commented:
/backport to release/7.0
